mllib/src/test/scala/org/apache/spark/ml/classification/FMClassifierSuite.scala

```diff
@@ -0,0 +1,147 @@
+
+import org.apache.spark.ml.classification.LogisticRegressionSuite.generateLogisticInput
+import org.apache.spark.ml.linalg._
+import org.apache.spark.ml.param.ParamsSuite
+import org.apache.spark.ml.regression.FMRegressorSuite._
+import org.apache.spark.ml.util._
+import org.apache.spark.ml.util.TestingUtils._
+import org.apache.spark.sql.{DataFrame, Row}
+
+class FMClassifierSuite extends MLTest with DefaultReadWriteTest {
+
+  import testImplicits._
+
+  private val seed = 42
+  @transient var smallBinaryDataset: DataFrame = _
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+
+    smallBinaryDataset = generateLogisticInput(1.0, 1.0, nPoints = 100, seed = seed).toDF()
+  }
+
+  test("params") {
+    ParamsSuite.checkParams(new FMClassifier)
+    val model = new FMClassificationModel("fmc_test", 0.0, Vectors.dense(0.0),
+      new DenseMatrix(1, 8, new Array[Double](8)))
+    ParamsSuite.checkParams(model)
+  }
+
+  test("FMClassifier: Predictor, Classifier methods") {
+    val sqlContext = smallBinaryDataset.sqlContext
+    import sqlContext.implicits._
+    val fm = new FMClassifier()
+
+    val model = fm.fit(smallBinaryDataset)
+    assert(model.numClasses === 2)
+    val numFeatures = smallBinaryDataset.select("features").first().getAs[Vector](0).size
+    assert(model.numFeatures === numFeatures)
+
+    testTransformer[(Double, Vector)](smallBinaryDataset.toDF(),
+      model, "rawPrediction", "probability", "prediction") {
+      case Row(raw: Vector, prob: Vector, pred: Double) =>
+        // Compare rawPrediction with probability
+        assert(raw.size === 2)
+        assert(prob.size === 2)
+        val probFromRaw1 = 1.0 / (1.0 + math.exp(-raw(1)))
+        assert(prob(1) ~== probFromRaw1 relTol 1E-6)
+        assert(prob(0) ~== 1.0 - probFromRaw1 relTol 1E-6)
+        // Compare prediction with probability
+        val predFromProb = prob.toArray.zipWithIndex.maxBy(_._1)._2
+        assert(pred == predFromProb)
+    }
+
+    ProbabilisticClassifierSuite.testPredictMethods[
+      Vector, FMClassificationModel](this, model, smallBinaryDataset)
+  }
+
+  test("factorization machines logisticLoss") {
+    // This testcase only tests whether the FM logloss part is valid and does not test the
+    // pairwise interaction logic. The pairwise interaction logic be tested in FMRegressor.
+    // When there is only one feature, FM will degenerate into LR. So two models will get
+    // almost same loss value.
+    def logLoss(modelRes: DataFrame): Double = {
+      modelRes.select("label", "probability").rdd.map {
+        case Row(label: Double, probability: DenseVector) =>
+          if (label > 0) -math.log(probability(1))
+          else -math.log(probability(0))
+      }.mean()
+    }
+
+    val fm = new FMClassifier().setMaxIter(200)
+    val fmModel = fm.fit(smallBinaryDataset)
+    val fmRes = fmModel.transform(smallBinaryDataset)
+    val fmLogLoss = logLoss(fmRes)
+
+    val lr = new LogisticRegression()
```

Comment:
I guess we do not need to compare FM with LR, check the convergen is enough.
I think we can more tests:
1, it seems that solver 'GD' is not tested at all;
2, it seemes that only dense datasets are used in testsuites;
3, add some cases like  'test("logistic regression doesn't fit intercept when fitIntercept is off")' / 'test("setThreshold, getThreshold")' / 'test("thresholds prediction")' / 'test("prediction on single instance")' in LogisticRegressionSuite